tests/usbredir_test.go

```diff
@@ -105,14 +107,34 @@ var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-c
 		It("Should fail when limit is reached", func() {
 			var tests [v1.UsbClientPassthroughMaxNumberOf + 1]testCase
 			for i := 0; i <= v1.UsbClientPassthroughMaxNumberOf; i++ {
-				ctx, cancelFn := context.WithCancel(context.Background())
-				tests[i] = testCase{
-					cancel: cancelFn,
-					err:    make(chan error),
+			retry_loop:
+				for try := 0; try < 3; try++ {
+					ctx, cancelFn := context.WithCancel(context.Background())
+					tests[i] = testCase{
+						cancel:  cancelFn,
+						connect: make(chan struct{}),
+						err:     make(chan error),
+					}
+					ctx = context.WithValue(ctx, "connected", tests[i].connect)
+					go runConnectGoroutine(virtClient, name, namespace, ctx, tests[i].err)
+
+					if i == v1.UsbClientPassthroughMaxNumberOf {
+						// Last test is meant to fail.
+						break
+					}
```

Comment:
With the retry loop, we are trying to handle errors related to the CI.
Shouldn't we try to cover these failures for the last connection?
We know that it should fail, and so it is ok, but we are also asserting that the problem will be 
`websocket: bad handshake`.
Maybe it is worth it to add a retry loop also for the last connection. WDYT?

Reply:
>  Maybe it is worth it to add a retry loop also for the last connection. WDYT?

The goal of the test is to have 4 well established usbredir connections so we can try the 5th and it'll fail. So, the retry is to make sure the environment is correct, that we have those 4 usbredir connections.

If we add the last one to the retry, it'll always fail till one slot is available by disconnecting one usbredir. That is handled after the retry loop.

Reply:
So, imho, feels like the logic order is fine this way
1. setup connections 5 connections
2. be sure that the first 4 is correct
3. test that the 5th fails

Reply:
Yes, everything is fine with the order.
But the 5th connection can fail for a different reason than `websocket: bad handshake`, no?
For example for the same reason that is causing the flakiness at the moment.

Reply:
>  For example for the same reason that is causing the flakiness at the moment.

The flakiness is weird but acceptable in that, we are failing read/write to TCP. I'm not sure why but I know this is somewhat new... expecting that to happen and redoing a retry is okay. I didn't forsee this, but it is indeed a possible reason for flake.

The 5th connection should fail earlier than write/read to the socket, because when we write/read the socket should already be promoted to a websocket. The bad handshake is the generic error to say we couldn't promote it to websocket. The actual reason is lost (that is a bug in itself) as it should show that maximum limit was reached, [here](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/rest/console.go#L112)

So, I think you are right that if there is another reason to failure of the 5th connection we should either retry or validate it, I just don't see it at this moment

Reply:
Thank you for the explanation!